idem_azurerm/states/azurerm/web/function_app.py

```diff
@@ -0,0 +1,571 @@
+      * ``secret``
+
+    Optional provider parameters:
+
+    **cloud_environment**: Used to point the cloud driver to different API endpoints, such as Azure GovCloud. Possible values:
+      * ``AZURE_PUBLIC_CLOUD`` (default)
+      * ``AZURE_CHINA_CLOUD``
+      * ``AZURE_US_GOV_CLOUD``
+      * ``AZURE_GERMAN_CLOUD``
+
+    Example acct setup for Azure Resource Manager authentication:
+
+    .. code-block:: yaml
+
+        azurerm:
+            default:
+                subscription_id: 3287abc8-f98a-c678-3bde-326766fd3617
+                tenant: ABCDEFAB-1234-ABCD-1234-ABCDEFABCDEF
+                client_id: ABCDEFAB-1234-ABCD-1234-ABCDEFABCDEF
+                secret: XXXXXXXXXXXXXXXXXXXXXXXX
+                cloud_environment: AZURE_PUBLIC_CLOUD
+            user_pass_auth:
+                subscription_id: 3287abc8-f98a-c678-3bde-326766fd3617
+                username: fletch
+                password: 123pass
+
+    The authentication parameters can also be passed as a dictionary of keyword arguments to the ``connection_auth``
+    parameter of each state, but this is not preferred and could be deprecated in the future.
+
+"""
+# Python libs
+from __future__ import absolute_import
+from dict_tools import differ
+import logging
+from datetime import datetime, timedelta
+import os
+
+# Azure libs
+HAS_LIBS = False
+try:
+    from azure.storage.blob import (
+        ResourceTypes,
+        AccountSasPermissions,
+        generate_account_sas,
+    )
+
+    HAS_LIBS = True
+except ImportError:
+    pass
+
+
+log = logging.getLogger(__name__)
+
+
+async def present(
+    hub,
+    ctx,
+    name,
+    resource_group,
+    functions_file_path,
+    os_type,
+    runtime_stack,
+    storage_account,
+    storage_rg=None,
+    app_service_plan=None,
+    functions_version=2,
+    enable_app_insights=None,
+    app_insights=None,
+    tags=None,
+    connection_auth=None,
+    **kwargs,
+):
+    """
+    .. versionadded:: 3.0.0
+
+    Ensure that a Function App exists.
+
+    :param name: The name of the Function App.
+
+    :param resource_group: The name of the resource group of the Function App.
+
+    :param functions_file_path: The file path of the compressed (zip) file containing any Azure Functions that should
+        be deployed to the Function App. The Azure Functions inside of this zip file should be using the same runtime
+        stack or language that is specified within the runtime_stack parameter. This file will be uploaded every
+        successfully run of this state. If there is a prior version of the zip file it will be overwritten. IMPORTANT:
+        The code for all the functions in a specific function app should be located in a root project folder that
+        contains a host configuration file and one or more subfolders. Each subfolder contains the code for a separate
+        function. The folder structure is shown in the representation below:
+            functionapp.zip
+             | - host.json
```

Comment:
what does this tree look like rendered? I've been putting my docstrings into Atom and viewing the RST to see what the docs will spit out. alternatively, I can add any test branches to readthedocs so we can preview them. let me know if you need that.